rpc: add new btcd extension command, setminingaddr

Roasbeef opened the pull request:
This commit adds a new `btcd` `JSON-RPC` extension command: `setminingaddr`.  This RPC allows the caller to dynamically add new mining addresses to `btcd'`s list of generation addresses. Before this RPC, doing such a thing would've required a full restart of `btcd`. 

The addition of this RPC also adds a new degree of flexibility to automated deployments of `btcd` using orchestration tools like `Docker`. The original inspiration for this RPC was spurred by some workarounds I had to implement when attempting to package `btcd` and `lnd` into a development environment using `docker-compose`.

A breakdown of the changes is as follows: 
 * A new RPC `setminingaddr` has been added. In order to ensure thread-safety, a mutex `miningAddrMTx` has been added to the `config` struct. This mutex must be held at all times when reading/modifying the `miningAddr` slice. 
 * The `cpuminer.Config` struct has been modified to use a closure function for obtaining new mining addresses. This decouple the current logic of selecting a mining address (randomly select from a list) from the internals of the `cpuminer` package. In the future if/when alternate selection logic is added, then code within the package shouldn't need to change at all. 

An interesting extension of this RPC was briefly discussed on IRC between myself and `davec`: the call could be extended to possibly accept a `BIP0032` extended public key rather than a series of addresses. Within the daemon, when generating new blocks, `btcd` could instead iterate through the HD keychain grabbing fresh addresses for each block rather than repeating through a statically coded list of addresses. Taking it a set further, further logic would be added to generate multi-sig `p2sh` addresses from addresses plucked from the keychain. 

For simplicity of this patch set, I've only implemented the most basic behavior round dynamically setting generation addresses, however the two extensions I mentioned above my be worth pursuing.

dajohi commented:
I'd rather have this be called addminingaddr since 'set' to me implies overwriting the current settings.  and then perhaps a delete as well?

Roasbeef commented:
Sure, what do you think about also throwing in a companion to the `add`, and `delete` calls: `list`?

stevenroose commented:
What happens when a node has multiple mining addresses?

Roasbeef commented:
@stevenroose the current algorithm pseudorandomly selects a mining address from the current set

davecgh commented:
I think having ``{add,delete}miningaddress` and `listminingaddresses` makes fine sense.  I haven't reviewed the code yet, but should be able to get it reviewed later this week.

Roasbeef commented:
I've pushed a new version of the PR which adds three new RPC's: `addminingaddr`, `delminingaddr` and `listminingaddrs`. I've also added a new integration tests to exercise the behavior and have also included a commit which fixes (what I consider) to be an existing bug in the configuration parsing.

davecgh commented:
Thanks for the updates.  One thing I noticed is the second commit converts the MiningAddr from `[]string` to `string` and I definitely don't think that should be changed.  There are already configurations out there that rely on the behavior it provides and changing it would break existing setups.

I know from IRC that you were seeing duplicates, but the reason for that is due to the following change:
```diff
-	preCfg := cfg
+	preCfg := &cfg
```

That was intentionally a copy. Making it a pointer causes it to set the same fields in the same config struct on the second parse which leads to appending entries in any slices twice.

davecgh commented on rpcserver.go:
```diff
@@ -3495,6 +3516,87 @@ func handleSetGenerate(s *rpcServer, cmd interface{}, closeChan <-chan struct{})
+	// If we've reached this point, then the targeted mining address wasn't
+	// found so we return an error indicating so.
+	return nil, &btcjson.RPCError{
+		Code:    btcjson.ErrRPCInternal.Code,
```
This isn't an internal error.  It was an invalid address provided by the caller.  Thus, it should be `btcjson.ErrRPCInvalidAddressOrKey`.

Roasbeef replied:
Fixed.

davecgh commented on rpcserverhelp.go:
```diff
@@ -488,6 +488,17 @@ var helpDescsEnUS = map[string]string{
 	"setgenerate-generate":     "Use true to enable generation, false to disable it",
 	"setgenerate-genproclimit": "The number of processors (cores) to limit generation to or -1 for default",
 
+	// AddMiningAddrCmd help.
```
Please alphabetize these too so it's all in the same order it's defined in the other file(s).

Roasbeef replied:
Fixed.

davecgh commented on rpcserverhelp.go:
```diff
@@ -488,6 +488,17 @@ var helpDescsEnUS = map[string]string{
+
+	// ListMiningAddrsCmd help.
+	"listminingaddrs--synopsis": "Returns a list of the current active mining addresses",
+
```
```Go
	"listminingaddrs--result0":  "List of current active mining addresses",
```

Otherwise:

```
$ btcctl help listminingaddrs
...
Result:
Nothing
```

Roasbeef replied:
Fixed.

davecgh commented on rpcserverhelp.go:
```diff
@@ -618,6 +629,9 @@ var rpcResultTypes = map[string][]interface{}{
 	"setgenerate":           nil,
+	"addminingaddr":         nil,
+	"delminingaddr":         nil,
+	"listminingaddrs":       nil,
```
This returns a slice of strings, so it should be:

```Go
{(*[]string)(nil)},
```

That would also trigger a test failure due to the missing `listminingaddrs--result0` help key I commented on above.

Roasbeef replied:
Fixed.

davecgh commented:
I forgot to mention in my review that I really like the idea here.  It will also simplify setting up simulation networks since you currently need btcd running first to get wallet to generate an address that you then have to stop btcd to configure with it.

It also allows miners to avoid downtime when changing things which is quite important for them.

Roasbeef commented on rpcserver.go:
```diff
@@ -169,6 +169,9 @@ var rpcHandlersBeforeInit = map[string]commandHandler{
 	"searchrawtransactions": handleSearchRawTransactions,
 	"sendrawtransaction":    handleSendRawTransaction,
 	"setgenerate":           handleSetGenerate,
+	"addminingaddr":         handleAddMiningAddr,
```
Fixed.

Roasbeef commented on rpcserver.go:
```diff
@@ -3495,6 +3516,87 @@ func handleSetGenerate(s *rpcServer, cmd interface{}, closeChan <-chan struct{})
 	return nil, nil
 }
 
+func handleAddMiningAddr(s *rpcServer, cmd interface{}, closeChan <-chan struct{}) (interface{}, error) {
```
Fixed.

Roasbeef commented on rpcserver.go:
```diff
@@ -3495,6 +3516,87 @@ func handleSetGenerate(s *rpcServer, cmd interface{}, closeChan <-chan struct{})
+	// Decode the addresses encoded as a string into an in-memory object.
+	addr, err := btcutil.DecodeAddress(c.Address, activeNetParams.Params)
+	if err != nil {
+		return nil, err
```
Fixed.

Roasbeef commented on rpcserver.go:
```diff
@@ -3495,6 +3516,87 @@ func handleSetGenerate(s *rpcServer, cmd interface{}, closeChan <-chan struct{})
+	if err != nil {
+		return nil, err
+	}
+
```
Nice catch! Fixed.

Roasbeef commented on rpcserver.go:
```diff
@@ -3495,6 +3516,87 @@ func handleSetGenerate(s *rpcServer, cmd interface{}, closeChan <-chan struct{})
+	return nil, nil
+}
+
+func handleDelMiningAddr(s *rpcServer, cmd interface{}, closeChan <-chan struct{}) (interface{}, error) {
```
Fixed.

Roasbeef commented on rpcserver.go:
```diff
@@ -3495,6 +3516,87 @@ func handleSetGenerate(s *rpcServer, cmd interface{}, closeChan <-chan struct{})
+	// Decode the addresses encoded as a string into an in-memory object.
+	addr, err := btcutil.DecodeAddress(c.Address, activeNetParams.Params)
+	if err != nil {
+		return nil, err
```
Fixed.

Roasbeef commented on rpcserver.go:
```diff
@@ -3495,6 +3516,87 @@ func handleSetGenerate(s *rpcServer, cmd interface{}, closeChan <-chan struct{})
+	if err != nil {
+		return nil, err
+	}
+
```
Fixed.

Roasbeef commented on rpcserver.go:
```diff
@@ -3495,6 +3516,87 @@ func handleSetGenerate(s *rpcServer, cmd interface{}, closeChan <-chan struct{})
+	}
+
+	cfg.miningAddrMtx.Lock()
+	for i := 0; i < len(cfg.miningAddrs); i++ {
```
Changed to use the single value range syntax instead. I always seems to forget about the semantics of the single value range syntax over slices!

Roasbeef commented on rpcserver.go:
```diff
@@ -3495,6 +3516,87 @@ func handleSetGenerate(s *rpcServer, cmd interface{}, closeChan <-chan struct{})
+
+	cfg.miningAddrMtx.Lock()
+	for i := 0; i < len(cfg.miningAddrs); i++ {
+		if cfg.miningAddrs[i].String() == addr.String() {
```
Hmm, yeah I agree with your analysis. Although I don't think any miners would be using p2pk address in this age, deletion should be more precise, hence the usage of `String` here.

Roasbeef commented on config.go:
```diff
@@ -133,7 +134,7 @@ type config struct {
 	MaxOrphanTxs         int           `long:"maxorphantx" description:"Max number of orphan transactions to keep in memory"`
 	Generate             bool          `long:"generate" description:"Generate (mine) bitcoins using the CPU"`
-	MiningAddrs          []string      `long:"miningaddr" description:"Add the specified payment address to the list of addresses to use for generated blocks -- At least one address is required if the generate option is set"`
+	MiningAddr           string        `long:"miningaddr" description:"Add the specified payment address to the list of addresses to use for generated blocks -- At least one address is required if the generate option is set"`
```
Created a new commit to implement this bug fix.

Roasbeef commented on config.go:
```diff
@@ -153,6 +154,7 @@ type config struct {
 	dial                 func(string, string) (net.Conn, error)
 	miningAddrs          []btcutil.Address
```
Fixed.

Roasbeef commented on config.go:
```diff
@@ -370,8 +372,8 @@ func loadConfig() (*config, []string, error) {
 	// the final parse below.
-	preCfg := cfg
-	preParser := newConfigParser(&preCfg, &serviceOpts, flags.HelpFlag)
+	preCfg := &cfg
```
Added

Roasbeef commented on btcjson/btcdextcmds.go:
```diff
@@ -89,6 +89,41 @@ func NewGetCurrentNetCmd() *GetCurrentNetCmd {
 }
 
+// AddMiningAddrCmd defines the setminingaddr JSON-RPC command.
+type AddMiningAddrCmd struct {
```
Fixed. Must've been left over from the name change from `setminingaddr` to `addminingaddr`.

Roasbeef commented on btcjson/btcdextcmds.go:
```diff
@@ -89,6 +89,41 @@ func NewGetCurrentNetCmd() *GetCurrentNetCmd {
+}
+
+// NewAddMiningAddrCmd returns a new instance which can be used to issue a
+// setminingaddr JSON-RPC command.
```
Fixed.

Roasbeef commented on mining/cpuminer/cpuminer.go:
```diff
@@ -334,9 +333,13 @@ out:
-		payToAddr := m.cfg.MiningAddrs[rand.Intn(len(m.cfg.MiningAddrs))]
+		// Obtain a new payment address selected according to the logic
+		// of the current mining address generation function.
+		payToAddr, err := m.cfg.GetMiningAddr()
```
Fixed.

Roasbeef commented on mining/cpuminer/cpuminer.go:
```diff
@@ -588,9 +591,13 @@ func (m *CPUMiner) GenerateNBlocks(n uint32) ([]*chainhash.Hash, error) {
-		payToAddr := m.cfg.MiningAddrs[rand.Intn(len(m.cfg.MiningAddrs))]
+		// Obtain a new payment address selected according to the logic
+		// of the current mining address generation function.
+		payToAddr, err := m.cfg.GetMiningAddr()
```
Fixed.

Roasbeef commented on integration/rpcserver_test.go:
```diff
@@ -94,10 +95,80 @@ func testGetBlockHash(r *rpctest.Harness, t *testing.T) {
+	// Next, try to delete a single address, then immediately try to delete
+	// the same address. The second call should result in an error as the
+	// address is no longer within the set of mining addresses.
+	if err := r.Node.DelMiningAddr(addrs[0]); err != nil {
```
Excellent, point. I've changed the duplicate deletion detection portion in the integration test in order to exercise deletion from near the mid-point of the slice.

Roasbeef commented on integration/rpcserver_test.go:
```diff
@@ -94,10 +95,80 @@ func testGetBlockHash(r *rpctest.Harness, t *testing.T) {
+		t.Fatalf("unable to delete mining address: %v", err)
+	}
+	if err := r.Node.DelMiningAddr(addrs[0]); err == nil {
+		t.Fatalf("deletion of non-existant address should fail")
```
Fixed.

Roasbeef commented:
Yep, after this PR portions of the initialization of the `rpctest.Harness` can also be re-written to utilize the RPC call rather than manually setting the flag on the command line.

davecgh commented:
So, I believe I know why you changed the `preCfg := cfg` copy to `preCfg := &cfg` originally which is the following vet error:
```
config.go:377::error: assignment copies lock value to preCfg: main.config contains sync.RWMutex (vet)
```

Unfortunately, as we've already discussed, it can't be a pointer because of the double parse.  Looks like the mutex is going to have to be refactored out of the config struct.  That could be a blessing as I suspect it will make the code a little cleaner anyways since the mutex won't have to be littered all throughout the code if it's encapsulated in a concurrent safe type.

Roasbeef commented:
I've just pushed a new version of the commit with the addressed review items squashed into place. I've also modified the implementation of the RPC's to utilize a new concurrent-safe struct which houses the mining address. As a result, the implementation is much cleaner as the mutex accesses are no longer strewn across the RPC calls. However, I wasn't quite sure where to put the new struct, so atm it lives in `server.go`.

davecgh commented on docs/json_rpc_api.md:
```diff
@@ -664,6 +667,44 @@ The following is an overview of the RPC methods which are implemented by btcd, b
+
+***
+
+***
```
Duplicate separator and extra blank line.

davecgh commented on docs/json_rpc_api.md:
```diff
@@ -664,6 +667,44 @@ The following is an overview of the RPC methods which are implemented by btcd, b
+|Method|listminingaddrs|
+|Parameters|None|
+|Description|Returns a list of all the currently active generation (coinbase payout) addresses.|
+|Returns|Nothing|
```
It returns the list of addresses, so this is not correct.

davecgh commented:
Nice job overall as usual!  Also, this needs a rebase to correct the conflicts.

davecgh commented:
Now the segwit has been merged, would you please rebase and update this to take care of the final review comments?  Thanks @Roasbeef!

jakesyl commented:
@jcvernaleo (as per #1530)

- Low priority
- Enhancement 

Should be closed immediately in favor of #1441
